File: python/data_structures/linked_list.py
"""
Singly Linked List Implementation
Time Complexity: O(1) for insertion at head, O(n) for others
Space Complexity: O(n)
"""

class Node:
    """Node class for linked list"""
    
    def __init__(self, data):
        """
        Initialize a node
        
        Args:
            data: Value to store in node
        """
        self.data = data
        self.next = None

class LinkedList:
    """Singly Linked List implementation"""
    
    def __init__(self):
        """Initialize empty linked list"""
        self.head = None
    
    def insert_at_beginning(self, value):
        """
        Insert at beginning of list
        
        Args:
            value: Value to insert
        """
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node
    
    def insert_at_end(self, value):
        """
        Insert at end of list
        
        Args:
            value: Value to insert
        """
        new_node = Node(value)
        
        if self.head is None:
            self.head = new_node
            return
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node
    
    def delete_node(self, value):
        """
        Delete node with given value
        
        Args:
            value: Value to delete
        """
        if self.head is None:
            return
        
        if self.head.data == value:
            self.head = self.head.next
            return
        
        temp = self.head
        while temp.next is not None and temp.next.data != value:
            temp = temp.next
        
        if temp.next is not None:
            temp.next = temp.next.next
    
    def search(self, value):
        """
        Search for a value in list
        
        Args:
            value: Value to search for
        
        Returns:
            True if found, False otherwise
        """
        temp = self.head
        while temp is not None:
            if temp.data == value:
                return True
            temp = temp.next
        return False
    
    def display(self):
        """Display linked list"""
        if self.head is None:
            print("List is empty")
            return
        
        temp = self.head
        elements = []
        while temp is not None:
            elements.append(str(temp.data))
            temp = temp.next
        print("List:", " -> ".join(elements) + " -> NULL")

if __name__ == "__main__":
    list = LinkedList()
    
    list.insert_at_end(10)
    list.insert_at_end(20)
    list.insert_at_end(30)
    list.insert_at_beginning(5)
    
    print("After insertions:")
    list.display()
    
    print(f"Searching for 20: {'Found' if list.search(20) else 'Not found'}")
    
    list.delete_node(20)
    print("After deleting 20:")
    list.display()

File: cpp/data_structures/linked_list.cpp
#include <iostream>
using namespace std;

// Singly Linked List Implementation
// Time Complexity: O(1) for insertion at head, O(n) for others
// Space Complexity: O(n)

struct Node {
    int data;
    Node* next;
    
    Node(int val) : data(val), next(nullptr) {}
};

class LinkedList {
private:
    Node* head;
    
public:
    LinkedList() : head(nullptr) {}
    
    // Insert at beginning
    void insertAtBeginning(int value) {
        Node* newNode = new Node(value);
        newNode->next = head;
        head = newNode;
    }
    
    // Insert at end
    void insertAtEnd(int value) {
        Node* newNode = new Node(value);
        
        if (head == nullptr) {
            head = newNode;
            return;
        }
        
        Node* temp = head;
        while (temp->next != nullptr) {
            temp = temp->next;
        }
        temp->next = newNode;
    }
    
    // Delete node with given value
    void deleteNode(int value) {
        if (head == nullptr) return;
        
        if (head->data == value) {
            Node* temp = head;
            head = head->next;
            delete temp;
            return;
        }
        
        Node* temp = head;
        while (temp->next != nullptr && temp->next->data != value) {
            temp = temp->next;
        }
        
        if (temp->next != nullptr) {
            Node* toDelete = temp->next;
            temp->next = temp->next->next;
            delete toDelete;
        }
    }
    
    // Search for a value
    bool search(int value) {
        Node* temp = head;
        while (temp != nullptr) {
            if (temp->data == value) return true;
            temp = temp->next;
        }
        return false;
    }
    
    // Display linked list
    void display() {
        if (head == nullptr) {
            cout << "List is empty" << endl;
            return;
        }
        
        Node* temp = head;
        cout << "List: ";
        while (temp != nullptr) {
            cout << temp->data << " -> ";
            temp = temp->next;
        }
        cout << "NULL" << endl;
    }
    
    // Destructor to free memory
    ~LinkedList() {
        Node* current = head;
        while (current != nullptr) {
            Node* next = current->next;
            delete current;
            current = next;
        }
    }
};

int main() {
    LinkedList list;
    
    list.insertAtEnd(10);
    list.insertAtEnd(20);
    list.insertAtEnd(30);
    list.insertAtBeginning(5);
    
    cout << "After insertions:" << endl;
    list.display();
    
    cout << "Searching for 20: " << (list.search(20) ? "Found" : "Not found") << endl;
    
    list.deleteNode(20);
    cout << "After deleting 20:" << endl;
    list.display();
    
    return 0;
}
